### ISSUE TITLE
Experimental parameterization issues for mask pre-training

### ISSUE THREAD

**Issue by ychuest** [qiaoqiaoLF/MxDNA#11]
I would like to ask you about the following parameter setting issues in pre-training. In your formal pre-training, how are these parameters set?
inputs_embeds: optionally [torrent.FloatTensor] = None, 
output_attentions: optionally [bool] = None, 
output_hidden_states: optionally [bool] = None, output_router_logits: optional [bool] = None, 
z_loss: optional [bool] = None,

class MxDNAForMaskedLM(MxDNAPreTrainedModel):
    _tied_weights_keys = ["lm_head.weight"]

    def __init__(self, config):
        super().__init__(config)
        self.model = MxDNAModel(config)
        self.vocab_size = config.vocab_size
        self.decoder = MxDNADecoderLayer(config, config.num_hidden_layers)
        self.lm_head = LMHead(config)
        self.num_experts = config.num_local_experts
        self.num_experts_per_tok = config.num_experts_per_tok
        self.norm = MxDNALayerNorm(config.hidden_size, eps=config.layer_norm_eps)
        self.router_aux_loss_coef = config.router_aux_loss_coef
        # Initialize weights and apply final processing
        self.post_init()

    def forward(
            self,
            input_ids: torch.LongTensor = None,
            attention_mask: Optional[torch.Tensor] = None,
            special_tokens_mask: Optional[torch.Tensor] = None,
            position_ids: Optional[torch.LongTensor] = None,
            inputs_embeds: Optional[torch.FloatTensor] = None,
            labels: Optional[torch.LongTensor] = None,
            output_attentions: Optional[bool] = None,
            output_hidden_states: Optional[bool] = None,
            output_router_logits: Optional[bool] = None,
            load_balancing_loss: Optional[bool] = None,
            z_loss: Optional[bool] = None,
    ) -> Union[Tuple, LanaguageModelingOutput]:
        r"""
        Args:
            labels (`torch.LongTensor` of shape `(batch_size, seq_len)`, *optional*):
                Labels for computing the masked language modeling loss. Indices should either be in `[0, ...,
                config.vocab_size]` or -100 (see `input_ids` docstring). Tokens with indices set to `-100` are ignored
                (masked), the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`.

        Returns:
        ```"""

        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions

        output_hidden_states = (
            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
        )
        output_router_logits = (
            output_router_logits if output_router_logits is not None else self.config.output_router_logits
        )
        batch_size, seq_len = input_ids.shape
        if position_ids is None:
            device = input_ids.device if input_ids is not None else inputs_embeds.device
            position_ids = torch.arange(
                0, seq_len, dtype=torch.long, device=device
            )
            position_ids = position_ids.unsqueeze(0).view(-1, seq_len)
        else:
            position_ids = position_ids.view(-1, seq_len).long()

        outputs = self.model(
            input_ids=input_ids,
            attention_mask=attention_mask,
            special_tokens_mask=special_tokens_mask,
            masked_tokens_mask=None,
            position_ids=position_ids,
            inputs_embeds=inputs_embeds,
            output_attentions=output_attentions,
            output_hidden_states=output_hidden_states,
            output_nucleotide_hidden_states=True,
            output_router_logits=output_router_logits,
        )

        motif_hidden_states = outputs.motif_hidden_states
        nucleotide_hidden_states = outputs.nucleotide_hidden_states
        motif_attention_mask = outputs.motif_attention_mask
        nucleotide_attention_mask = attention_mask

        decoder_layer_outputs = self.decoder(
            hidden_states_q=nucleotide_hidden_states,
            hidden_states_kv=motif_hidden_states,
            attention_mask_q=nucleotide_attention_mask,
            attention_mask_kv=motif_attention_mask,
            position_ids=position_ids,
            output_attentions=output_attentions,
        )

        last_hidden_states = decoder_layer_outputs[0]
        last_hidden_states = self.norm(last_hidden_states)

        logits = self.lm_head(last_hidden_states)
        logits = logits.float()

        loss = None
        if labels is not None:
            # No shift for masked language modeling
            mlm_logits = logits[..., :, :].contiguous()
            mlm_labels = labels[..., :].contiguous()
            # Flatten the tokens
            loss_fct = CrossEntropyLoss()
            mlm_logits = mlm_logits.view(-1, self.config.vocab_size)
            mlm_labels = mlm_labels.view(-1)
            # Enable model parallelism
            mlm_labels = mlm_labels.to(mlm_logits.device)
            loss = loss_fct(mlm_logits, mlm_labels)

        aux_loss = None
        router_z_loss = None
        if output_router_logits and load_balancing_loss:
            aux_loss = load_balancing_loss_func(
                outputs.router_logits_list,
                outputs.motif_mask_center_list, )

            if labels is not None:
                loss += self.router_aux_loss_coef * aux_loss.to(loss.device)  # make sure to reside in the same device
        if output_router_logits and z_loss:
            router_z_loss = router_z_loss_func(
                outputs.router_logits_list,
                outputs.motif_mask_center_list, )
            loss += self.router_aux_loss_coef * router_z_loss.to(loss.device)

        return LanaguageModelingOutput(
            loss=loss,
            aux_loss=aux_loss,
            z_loss=router_z_loss,
            logits=logits,
            router_logits=outputs.router_logits_list,
            all_hidden_states=outputs.all_hidden_states + decoder_layer_outputs[0] if output_hidden_states else None,
            all_attentions=outputs.all_attentions + decoder_layer_outputs[1] if output_attentions else None,
        )

**Comment by qiaoqiaoLF** [comment]
Thanks for raising this issue and for your interest in the pre-training details!

Regarding the parameters you mentioned: `inputs_embeds`, `output_attentions`, `output_hidden_states`, `output_router_logits`, and `z_loss`, these arguments are currently **not directly utilized** in our pre-training process (But you can still give it a try) .

The specific configurations and settings we employed for pre-training are detailed in the **Appendix** of the paper. Please refer to that section for a comprehensive overview of the hyperparameters and training setup we used.

Let me know if you have any further questions after reviewing the Appendix.
